src/eth/TransactionManager.js

```diff
@@ -11,13 +14,93 @@ export default class TransactionManager extends PublicService {
+  parseContractFunctionArgs(args) {
+    let metadata = null;
+    let dsProxyAddress = null;
+
+    if (
+      typeof args !== 'undefined' &&
+      Array.isArray(args) &&
+      typeof args[args.length - 1] === 'object'
+    ) {
+      // Detect additional metadata attatched to last arg and merge it with default metadata
+      if (args[args.length - 1].hasOwnProperty('metadata')) {
+        metadata = {};
+        Object.assign(metadata, args[args.length - 1].metadata);
+        delete args[args.length - 1].metadata;
+      }
+      // Detect proxy address and route through DSProxy contract
+      if (args[args.length - 1].hasOwnProperty('dsProxyAddress')) {
+        dsProxyAddress = args[args.length - 1].dsProxyAddress;
+        delete args[args.length - 1].dsProxyAddress;
+      }
+      // If last arg item is an empty object, remove it
+      if (Object.keys(args[args.length - 1]).length === 0) args.pop();
+    }
+    return { additionalMetadata: metadata, dsProxyAddress };
+  }
+
   formatHybridTx(contract, key, args, name, businessObject = null) {
+    // Strip the key to just the method name, in cases where the method was called
+    // using the full method sig e.g. contract["draw(address,uint256)"](foo, bar)
+    const method = key.replace(/\(.*\)$/g, '');
+
+    const metadata = { contract: name, method, args };
+    const {
+      additionalMetadata,
+      dsProxyAddress
+    } = this.parseContractFunctionArgs(args);
+
+    // Handle any additional metadata
+    if (additionalMetadata !== null) {
+      this.get('log').debug(
+        'Attaching additional tx metadata:',
+        additionalMetadata
+      );
+      Object.assign(metadata, additionalMetadata);
+    }
+
+    // DSProxy handling – different from the fact that this is a Proxy class ;)
+    if (dsProxyAddress !== null) {
+      this.get('log').debug(
+        'Calling ' + key + ' via DSProxy at ' + dsProxyAddress
+      );
+      const dsProxyContract = wrapContract(
+        new Contract(
+          dsProxyAddress,
+          dappHub.dsProxy,
+          this.get('web3')
+            .ethersProvider()
+            .getSigner()
+        ),
+        'DSProxy',
+        dappHub.dsProxy,
+        this
+      );
+      // Pass in any additional tx options passed to this tx (e.g. value, gasLimit)
+      // if the last arg is an object literal (not a BigNumber object etc.)
+      let options = {};
+      if (
+        typeof args[args.length - 1] === 'object' &&
+        args[args.length - 1].constructor === Object
+      ) {
+        Object.assign(options, args[args.length - 1]);
+        args.pop();
+      }
+      // Assign proxied tx metadata and options to proxy tx
+      Object.assign(options, { metadata });
+      // Get proxied tx calldata to pass to DSProxy
+      const data = contract.interface.functions[key](...args).data;
+      return dsProxyContract.execute(contract.address, data, options);
```

Comment:
since you're returning here, doesn't this mean that proxy calls don't get the nonce injected into their arguments and they don't get wrapped in the hybrid object with lifecycle hooks?

i think there's a refactoring that should happen here so we treat all calls the same, something like:
```js
const contractCall = argsHasDsProxyAddress()
  ? createProxyCall()
  : newArgs => contract[key](...newArgs);

return this.createHybridTx(
  this.injectSettings(args).then(contractCall), 
  { 
    businessObject, 
    metadata: combinedMetadata(name, method, args) 
  }
);
```

Reply:
Oh, I get it now, I'm mistaken, because you're using `wrapContract` again. Hmm.

Reply:
Those additional args (like `nonce`) get merged into `options`, and because `dsProxyContract.execute` calls the contract method directly it’ll be intercepted by the `Proxy` class and run through that function again (except this time without the `dsProxyAddress` option set, so it won’t pass into that `if` block).